### ISSUE TITLE
Blockwise memory exceeded when creating random array

### ISSUE THREAD

**Issue by TomNicholas** [tomwhite/cubed#160]
I don't understand why this raises an error:

I had thought that the max_mem was referring to the total amount of data that had to be in RAM at any one time, which here should presumbly just be one chunk, i.e. 

If alternatively the max_mem refers to the total size of the array, that still doesn't make sense to me in this case, because the total size of the array is X, not Y.

**Comment by TomNicholas** [comment]
(I want to use `cubed.random.random` as the first step in some simple benchmarks against dask/ramba)

**Comment by tomwhite** [comment]
There is effectively a 4x overhead on the chunk size, as explained here: https://github.com/tomwhite/cubed/blob/74145fb826d21496ca86b4e64b0fcb1c55d8a931/cubed/primitive/blockwise.py#L192-L205.

What you can do is increase the `max_mem` and then see how much was actually used for a job by using a [HistoryCallback](https://github.com/tomwhite/cubed/blob/74145fb826d21496ca86b4e64b0fcb1c55d8a931/cubed/extensions/history.py#L11) which will save a CSV locally with memory stats. This might help you then reduce it again.

> (I want to use `cubed.random.random` as the first step in some simple benchmarks against dask/ramba, in particular to try out [pangeo-data/distributed-array-examples#2](https://github.com/pangeo-data/distributed-array-examples/issues/2))

Nice!

BTW I did a couple of pangeo-style examples which might be helpful: https://github.com/tomwhite/cubed/blob/main/examples/pangeo-vorticity.ipynb, and https://github.com/tomwhite/cubed/blob/main/examples/pangeo-tem.ipynb

**Comment by TomNicholas** [comment]
Thanks! 

So I just ran this successfully using xarray! [Notebook here](https://gist.github.com/TomNicholas/8366c917349b647d87860a20a257a3fb)

> What you can do is increase the max_mem and then see how much was actually used for a job by using a [HistoryCallback](https://github.com/tomwhite/cubed/blob/74145fb826d21496ca86b4e64b0fcb1c55d8a931/cubed/extensions/history.py#L11) which will save a CSV locally with memory stats. This might help you then reduce it again.

I did that and it looks like the actual memory used was just over 1.5GB, which makes sense I think. 

It took about 33 minutes to execute, but I didn't specify an executor so presumably that means it executed just using a single python process? So for a full comparison against dask.distributed what executor should I use to get parallelism?

---

As an aside I find it a bit confusing that `max_mem` sometimes means "maximum memory the computation is allowed to use" (i.e. actual usage) and sometimes means "maximum memory the computation is projected to use" (i.e. a conservative estimate). For example the DAG visualization just shows me the number I entered as `max_mem`, which isn't particularly helpful here. It would be more helpful if it showed two numbers, i.e. `projected_memory_use / max_allowed_mem`, so I can see if I will exceed it or how much headroom I have.

![image](https://user-images.githubusercontent.com/35968931/236066767-c82bc76a-93fc-4d76-8844-9f284c05baa4.png)

**Comment by tomwhite** [comment]
Awesome!

> It took about 33 minutes to execute, but I didn't specify an executor so presumably that means it executed just using a single python process? So for a full comparison against dask.distributed what executor should I use to get parallelism?

By default tasks are run sequentially using the local Python process (using the PythonDagExecutor). For parallelism have a look at Lithops or Modal, see https://github.com/tomwhite/cubed/blob/main/examples/README.md. 

> As an aside I find it a bit confusing that `max_mem` sometimes means "maximum memory the computation is allowed to use" (i.e. actual usage) and sometimes means "maximum memory the computation is projected to use" (i.e. a conservative estimate).

That's a good point. Using terms like "projected" and "actual" might be a good change to make to clarify what the memory values mean. I tried to draw a diagram of how to think about memory here: https://tom-e-white.com/cubed/computation.html#memory - is "required_mem" what you were thinking of as "projected"?

**Comment by TomNicholas** [comment]
I updated the notebook gist - using lithops (on my laptop!) It took 48 seconds. That's awesome, I'm very impressed and excited by this performance!

Yes required_mem is what I meant by "projected_mem". That docs page is helpful, but the UI conflates these concepts a bit. For example the saved history records a variable "required_mem", but sometimes that means required_mem and sometimes means max_mem. I already know max_mem because I specified that, if I'm looking at the history I'm doing it because I want to know required_mem. (If these were kept distinct I wouldn't have needed to drop the rows from the data frame where required_mem=max_mem in my notebook.)

**Comment by tomwhite** [comment]
I had another look at this and noticed that you are not picking up the actual, measured memory used by tasks. Have a look at this notebook for how to do that with the local Lithops executor (it has to run tasks in separate processes to be able to measure task memory properly):

https://github.com/tomwhite/cubed/blob/main/examples/lithops-add-random-local.ipynb

You can see the (maximum) memory utilization calculation at the bottom, which calculates actual (peak) memory used divided by projected ("required") memory. The idea is that we want to keep the utilization under 100%.

Sorry I didn't point this out before!

**Comment by tomwhite** [comment]
@TomNicholas I ran the quadratic benchmark using lithops locally and generated the analysis chart for peak memory usage and memory utilization here:

https://github.com/tomwhite/cubed/blob/benchmark-quadratic-peak-mem/examples/benchmark-quadratic.ipynb

Peak was a little under 2.2GB, and utilization was <1.06 (need to have a look at why it was over 1 - that could indicate a bug in Cubed).

(We should make it easier to get this kind of analysis from a computation.)

**Comment by TomNicholas** [comment]
Thanks for this clarification!

> the local Lithops executor (it has to run tasks in separate processes to be able to measure task memory properly):

How are you telling Lithops to do this? Through the `LITHOPS_LOCAL_CONFIG = {"lithops": {"backend": "localhost", "storage": "localhost"}}` setting?

**Comment by tomwhite** [comment]
> How are you telling Lithops to do this? Through the `LITHOPS_LOCAL_CONFIG = {"lithops": {"backend": "localhost", "storage": "localhost"}}` setting?

Yes. That makes sure that you use localhost, overriding any settings that you may have on your machine.
